### PR TITLE
Kuzmin Turek implementation for advection with constant velocity

### PR DESCRIPTION
**Pull request by WilkAndy**
Please comment and advise on how to do this properly, @joshuahansel , @permcody , @andrsd 

a

### CODE REVIEW

**Comment by joshuahansel** [comment]
I still don't think FCT is even possible until someone comes up with a way to do multiple solves in a time step, since the antidiffusive fluxes generally depend on the high-order solution. An exception would be if both of the following are true:
1. The high-order scheme uses a lumped mass matrix.
2. Fully explicit time integration is used.

In this case, the anti-diffusive fluxes would only depend on the old solution. I don't know how great the overall scheme would be if 1. is satisfied.

**Comment by WilkAndy** [comment]
OK, thanks for your comments on FCT, @joshuahansel .  Today i want to do a worked example (by hand) that will hopefully convince me of whether FCT is going to be possible and worthwhile.

**Comment by WilkAndy** [comment]
I believe that Kuzmin-Turek works.  I'm really excited.  There's a lot of work to do, probably both in the framework and PorousFlow, but i'm attaching a figure illustrating the exciting results.  I hope that @hsheldon is as please as i am, and if we get this working for the general case, i think it'll be an extremely well-received addition.



@joshuahansel , i don't think two separate solves are needed.  Instead steps (1) to (3) in KT Fig 2 just modify the advection Kernel, by adding diffusion (D) and anti-diffusion (F).  This is what is implemented in the attached PR.  Because of the stuff below, i have only been able to implement KT in 1D for constant velocity where i can compute their "K" matrix by hand.

@andrsd , @permcody , @joshuahansel , here's what i think we need to make this general:
 - Ability to compute KT's K matrix, given in their Eqn(20).  This looks very much like our usual residuals, save for lack of multiplication by _u[_qp] in the integral, so it can almost be computed by a Kernel.  However, the key is that k_ij must be computed for all linked pairs of nodes, and be stored somewhere so that the remainder of the algorithm can use it.  We can certainly compute it by looping over elements, and just increment the appropriate k_ij entries.  Almost certainly, the velocity, v, will be a MaterialProperty in PorousFlow, so it'll have to go into the integral.  We will also need derivatives of k_ij wrt to all the variables.
 - Ability to get a list of all the nodes connected to a certain node.  These appear in the sums over "j" in KT.
 - Ability for the Kernel to access the k_ij, given global node numbers of the element (or similar), as well as nodal values of u.  Probably u will be a "nodal" Material, like many other things in PorousFlow.

I'd like your help with this, please.
fltvd_comparison

**Comment by WilkAndy** [comment]
I forgot to mention, the figure is generated in the same way as the final figure in the recent numerical_diffusion doco (http://mooseframework.inl.gov/docs/PRs/12348/site/modules/porous_flow/numerical_diffusion.html)

**Comment by joshuahansel** [comment]
Those results look great. It appears that the algorithm that you and KT are using is the different from the FCT algorithm I know. I've read the paper a little, but I think it would take a longer time investment for me to truly understand it all. Specifically, the issue of confusion is the computation of the anti-diffusive fluxes. As I had known it, the anti-diffusive fluxes, when applied without limitation, were supposed to reproduce the high-order solution - this property yielded the scheme that I had been using. It appears in KT's paper that the anti-diffusive fluxes were not designed around this property, and thus they did not need to depend on the high-order solution (I guess they just depend upon the implicit solution?).

I think your questions at this point are MOOSE questions, so I'll leave that to the MOOSE experts, as I need to spend my time working on RELAP-7. I'd be very much interested to see what can be done further with this work though! Clearly, I need to read some more on FCT sometime...

**Comment by andrsd** [comment]
> Ability to compute KT's K matrix, given in their Eqn(20).  This looks very much like our usual residuals, save for lack of multiplication by _u[_qp] in the integral, so it can almost be computed by a Kernel.  However, the key is that k_ij must be computed for all linked pairs of nodes, and be stored somewhere so that the remainder of the algorithm can use it.  We can certainly compute it by looping over elements, and just increment the appropriate k_ij entries.

Sounds like a job for `UserObject`. You should be able to get all the linked nodes via libMesh.  For serial runs this will be easy. When you get to implement the parallel version, you will probably need to use the relationship manager @permcody mentioned earlier, so that you can get the right level of ghosting.

> Ability to get a list of all the nodes connected to a certain node. These appear in the sums over "j" in KT.

This sounds related to (1) - it should not be a problem

> Ability for the Kernel to access the k_ij, given global node numbers of the element (or similar), as well as nodal values of u.

You can design your UserObject (as I mentioned above) such that you can request the data the way you need them. 

Getting nodal values of variables has been in the framework for a long time - this should not a be a problem - look up `coupledNodalValue` in `Coupleable` interface.

> Probably u will be a "nodal" Material, like many other things in PorousFlow.

I am not sure I understand this. We solve for `u`, right? So why nodal material?

And just out of curiosity, you want KT because you want to stick with continuous Galerkin?

**Comment by WilkAndy** [comment]
> Sounds like a job for `UserObject`. You should be able to get all the linked nodes via libMesh. For serial runs this will be easy. When you get to implement the parallel version, you will probably need to use the relationship manager @permcody mentioned earlier, so that you can get the right level of ghosting.

Yes, i used `UserObject` in the current implementation, but i didn't know i could get the linked nodes via libMesh (i just assumed the dofs at the nodes were ordered from left-to-right along my 1D line).  Could you point me to some methods, please?  Libmesh is quite opaque to me.

> 
> > Ability to get a list of all the nodes connected to a certain node. These appear in the sums over "j" in KT.
> 
> This sounds related to (1) - it should not be a problem
> 
> > Ability for the Kernel to access the k_ij, given global node numbers of the element (or similar), as well as nodal values of u.
> 
> You can design your UserObject (as I mentioned above) such that you can request the data the way you need them.
> 
> Getting nodal values of variables has been in the framework for a long time - this should not a be a problem - look up `coupledNodalValue` in `Coupleable` interface.

I had to screw around a little in the current implementation, probably because of the lack of knowing what the neighbouring nodes are.  My `Kernel` will ask: "what is the flux correction for node such-and-such" and my `UserObject` will then need to compile a list of neighbouring nodes, pull out K_ij and the value of "u" at those nodes, compute and return the final result.

I couldn't do that with coupledNodalValue because of not knowing what the neighbours were.

> 
> > Probably u will be a "nodal" Material, like many other things in PorousFlow.
> 
> I am not sure I understand this. We solve for `u`, right? So why nodal material?
> 

This is just because of the architecture of PorousFlow: most things are nodal Materials.  But if you like you can think of "u" as a nonlinear function of the `Variables` of the problem: it is exactly the same as a nodal Material, but in a different guise.

> And just out of curiosity, you want KT because you want to stick with continuous Galerkin?

Yes, primarily.  In many parts of PorousFlow we assume that the Variables are linear Lagrange.  (Actually, the KT idea may allow us to get around a lot of this.)  RDG clashes with a lot of PorousFlow and we'll need a lot of subtle changes to integrate those ideas into PorousFlow.  I suspect that KT and RDG(P0P1) yield very similar results.

**Comment by WilkAndy** [comment]
@joshuahansel - yes, the antidiffusive fluxes are cleverly based on how much heat is moving to this node and how much is moving away.  So you don't need to solve anything before you compute the antidiffusion - you just base your computation on the current solution!   In 1D it's doing *exactly* the same as RDG(P0P1), but swap nodes with elements.  In higher dimensions it's probably the same too, but i haven't checked by hand.

**Comment by WilkAndy** [comment]
Don't worry, @nadrsd , google helped me.  I found this thread, which i presume is still correct:

https://sourceforge.net/p/libmesh/mailman/message/34587502/

a

**Comment by WilkAndy** [comment]
Hi @andrsd .  So i played around with this for a few hours today.  i made a little progress utilizing the neighbouring nodes, thanks.

In my `GeneralUserObject` i have a loop over elements:
```
 for (const auto & elem : _mesh.getMesh().active_element_ptr_range())
```
Within this loop, i'd like to do this sort of thing:
```
  for (unsigned i = 0; i < num_nodes; ++i)                                                                                                                                                                          
    for (unsigned j = 0; j < num_nodes; ++j)                                                                 
      for (unsigned qp = 0; qp < _qrule->n_points(); ++qp)                                                   
        kk[i][j] += _JxW[qp] * _coord[qp] * (_grad_test[i][qp] * _velocity) * _test[j][qp];  
```
but `GeneralUserObject` doesn't know `_qrule`, `_JxW`, `_coord`, `_grad_test` or `_test`.  What's a good way for `GeneralUserObject` to access these things ?

**Comment by andrsd** [comment]
> `GeneralUserObject` doesn't know `_qrule`, `_JxW`, `_coord`, `_grad_test` or `_test`

Yeah, because it is not associated with any mesh entity. I made a special mesh sweeping class that derives from `GeneralUserObject` to accommodate the rDG needs - it is in the rDG module and is called `ElementLoopUserObject`. That will give you `_JxW`, `_coord` and `_q_rule`, but not `_test` or `_grad_test`, because we did not need them. But, it provides bunch of virtuals that you can use to do what you need.  The one you want to play with are `computeElement`, `computeBoundary`, `computeInternalSide`. But for your case, you probably want to put some code inside `computeElement`.  You will need to this as first:
```
_fe_problem.prepare(_current_elem, _tid);
_fe_problem.reinitElem(_current_elem, _tid);
```
that will interpolate the variable values and reinit test functions, gradients, etc.

If your user object acts on a variable, you will probably want to use this:
```
    MooseVariableInterface<Real>(this,
                                 false,
                                 "variable",
                                 Moose::VarKindType::VAR_NONLINEAR,
                                 Moose::VarFieldType::VAR_FIELD_STANDARD),
    _var(*mooseVariable()),
    _test(_var.phi()),
    _grad_test(_var.gradPhi()),
    _phi(_assembly.phi(_var)),
    _grad_phi(_assembly.gradPhi(_var)),
```
That is from `Kernel.C` (go look there for more details). That will give you the usual stuff you are used to.  If you do not have a variable, there is another way to get `_test` and `_grad_test`:
```
    _phi(_assembly.fePhi<Real>(_fe_type)),
    _grad_phi(_assembly.feGradPhi<Real>(_fe_type)),
```
This is tailored to your specific problem - `Real` means scalar-valued FE-type and for `_fe_type`, you might have something like `(FIRST, LAGRANGE)`.

If you go with the `MooseVariable` approach, you will get one sweep through mesh for each instance of that user object and that's probably not what you want when you move to the implementation for porous flow. You will want one sweep and take care of all the caching at once. So you need to decide how to get the `_fe_type` in the user object.  One way might be, that you provide a list of variables you will be operating on, you check that all have the same `FEType` and that would be the one you use. I do not think you want to combine `FEType`s...

**Comment by andrsd** [comment]
> can you give me some idea of how to get all nodes attache to a certain node?

There is also `MooseMesh::nodeToElemMap()` that will give you all elements connected to a node. Then you can loop over nodes of each element skipping the one you passed into the `nodeToElemMap` call...

**Comment by WilkAndy** [comment]
Thanks for your answers about `_JxW`, etc, @andrsd .   I think i've got the whole day to work on this!

**Comment by WilkAndy** [comment]
Hi @andrsd , @permcody , @joshuahansel .   I think this PR would benefit from a brief review.  You're probably short on time, so please concentrate your efforts on the `AdvectiveFluxCalculator` `UserObject`, as the remainder of the stuff is either doco or pretty straightforward Kernel stuff.

**Comment by WilkAndy** [comment]
I've realised that in multiple dimensions, the Kernel must divide it's residual contributions by the valence of each edge.  Viz, if an edge between two nodes is shared by N elements, then the Kernel's contribution must be divided by N.  The current Kernel implementation loops over all elements, and adds the residual contributions N times.  All i need do is to divide K_ij by the valence of the i-j edge.

A faster and cleaner way would be for the Kernel to loop over nodes, computing the residual contributions directly from all connected nodes (using K_ij).  Probably not very MOOSian, however.

**Comment by andrsd** [comment]
BTW @WilkAndy : why did you use `GeneralUserObject`? `ElementUserObject` would have all the `_JxW`, `_coord`, variable values, etc. for you.

File: modules/porous_flow/src/userobjects/AdvectiveFluxCalculator.C
Comment (andrsd) [inline modules/porous_flow/src/userobjects/AdvectiveFluxCalculator.C]: Yes, this is already in MooseMesh, so use that instead of solving that problem again

File: modules/porous_flow/src/userobjects/AdvectiveFluxCalculator.C
Comment (andrsd) [inline modules/porous_flow/src/userobjects/AdvectiveFluxCalculator.C]: Why not `node_i->id()`? You are using the whole `Node` class as an index + doing a memory copy here.

File: modules/porous_flow/include/userobjects/AdvectiveFluxCalculator.h
Comment (andrsd) [inline modules/porous_flow/include/userobjects/AdvectiveFluxCalculator.h]: Use `ElementUserObject` if possible. Not sure you have any limitations that would prevent it. From looking at your code briefly, it should be doable.

Comment (andrsd) [reply modules/porous_flow/include/userobjects/AdvectiveFluxCalculator.h]: You should also be able to compute the edge valency in this user object and then you would pull it out into the kernel via your defined API.

File: modules/porous_flow/include/userobjects/AdvectiveFluxCalculator.h
Comment (andrsd) [inline modules/porous_flow/include/userobjects/AdvectiveFluxCalculator.h]: If use decide to try the `ElementUserObject`, you would have `coupledNodalValue` API available which would be more MOOSEian.

File: modules/porous_flow/src/kernels/FluxLimitedTVDAdvection.C
Comment (andrsd) [inline modules/porous_flow/src/kernels/FluxLimitedTVDAdvection.C]: Useless line 🤪

File: modules/porous_flow/src/userobjects/AdvectiveFluxCalculator.C
Comment (andrsd) [inline modules/porous_flow/src/userobjects/AdvectiveFluxCalculator.C]: Why would not you just traverse through `_kij` and set its elements to zero? Are you trying to preallocate the data structure? Because you could build up the data structure on the first pass and the subsequent ones would just zero it out. Just an idea.

Comment (andrsd) [reply modules/porous_flow/src/userobjects/AdvectiveFluxCalculator.C]: If you want to stick with this though, look at `MooseMesh::localNodesBegin` and `MooseMesh::localNodesEnd`.

File: modules/porous_flow/src/userobjects/AdvectiveFluxCalculator.C
Comment (WilkAndy) [reply modules/porous_flow/src/userobjects/AdvectiveFluxCalculator.C]: Thanks for all your comments, @andrsd .    For this one (use `node_i->id()`), are you suggesting my _kij map should be
```
std::map<dof_id_type, std::map<dof_id_type, Real>>
```
instead of `std::map<Node, std::map<Node, Real>>` ?

File: modules/porous_flow/src/userobjects/AdvectiveFluxCalculator.C
Comment (WilkAndy) [reply modules/porous_flow/src/userobjects/AdvectiveFluxCalculator.C]: I can use `nodeToElemMap` (which returns a `std::map< dof_id_type, std::vector< dof_id_type > >` )but then i can't see how i can use `MeshTools::find_nodal_neighbors` since the latter requires a `std::vector<std::vector<const Elem *>>`.  I don't really care about finding neighbors next to nodes: i just want to find the nodal neighbors, but the elements seem to be a necessary intermediate step

**Comment by permcody** [comment]
Look on libMesh’s Mesh or MeshBase object. There are methods that will give
you a Node or Elem object if you have the ID. No searching required!

**Comment by WilkAndy** [comment]
Thanks @permcody .   I see  `virtual const Node& libMesh::MeshBase::node_ref  (  const dof_id_type  i  ) ` which should do the trick.
In writing this code, i've been a bit worried about all this node and element numbering.  Eg, how do i tell that `i` in the above method is actually equivalent to `node.id()`.  Maybe it's equivalent to `node.unique_id()` or one of the other types of "id".

File: modules/porous_flow/src/userobjects/AdvectiveFluxCalculator.C
Comment (dschwen) [reply modules/porous_flow/src/userobjects/AdvectiveFluxCalculator.C]: I wonder if `std::map<std::pair<dof_id_type, dof_id_type>, Real>` would be even better here.

File: modules/porous_flow/src/userobjects/AdvectiveFluxCalculator.C
Comment (andrsd) [reply modules/porous_flow/src/userobjects/AdvectiveFluxCalculator.C]: I misunderstood line 74 and 75 below. I thought you were trying to build what we already have in MooseMesh. On second look, it is not the case, so your code is fine.

But, you will want to change the local variable name from `_nodes_to_elem_map` to `nodes_to_elem_map` :-)

File: modules/porous_flow/src/userobjects/AdvectiveFluxCalculator.C
Comment (WilkAndy) [reply modules/porous_flow/src/userobjects/AdvectiveFluxCalculator.C]: Decided against the std::pair idea.  I now use _kij to figure out the connected nodes, so the `std::map<dof_id_type, std::map<dof_id_type, Real>>` is better.  But i'm going to change _valence to use the std::pair.

File: modules/porous_flow/include/userobjects/AdvectiveFluxCalculator.h
Comment (WilkAndy) [reply modules/porous_flow/include/userobjects/AdvectiveFluxCalculator.h]: I'm not sure why you're proposing `ElementUserObject`, @andrsd .  I want to do a loop over elements, computing my `_kij` and your `ElementLoopUserObject`, with its `pre()` and `computeElement()` methods seem to be exactly what i want.

**Comment by WilkAndy** [comment]
I'll be away for much of November, so i've asked @hsheldon whether she might like to add some tests and/or doco here.  In no particular order, this is what i would like to do:

- [ ] Test that the scheme works on different meshes.  When i say "the scheme works", i mean, observing that the code just doesn't crash, and attempting to see that the advection speed is as desired.  The meshes include
  - 2D and 3D meshes as well as the 1D i've explored already
  - meshes with different blocks
  - meshes with different blocks where the "u" is only defined on a subset of those blocks
  - meshes with not just quad elements, but some triangles too

 - [ ] Test that the scheme works with different number of processors.  I just want to know whether the results are the same as the single-proc case, because all this screwing around with MOOSE at a lower level than usual might have messed something up.

 - [ ]  Test that the scheme works with mesh adaptivity.

 - [ ] Test that the scheme works with multiple threads.

 - [ ]  Add the full Jacobian stuff that is computed in `_dflux_out_dvar` in `FluxLimitedTVDAdvection::computeJacobian()`.  At present only the contributions local to the current element are added.  Verify (hope) that these contributions are correct in some tests.

 - [ ]  Add another 2 cases to the numerical_diffusion documentation (http://mooseframework.org/moose/modules/porous_flow/numerical_diffusion.html) one with flux_limiter_type=None (which should be identical to the full-upwind case) and one with flux_limiter_type=superbee (or whatever) that should be identical to the RDG(P0P1) case.

 - [ ]  Complete the kt_worked.md documentation

 - [ ]  Document (in a separate page) the effect of different flux_limiter_types, in a similar way to the example on the RDG page: http://mooseframework.org/moose/modules/rdg/index.html

**Comment by WilkAndy** [comment]
@hsheldon , the test failures are wierdo ones that i think aren't related to this PR

File: modules/porous_flow/include/userobjects/AdvectiveFluxCalculator.h
Comment (andrsd) [reply modules/porous_flow/include/userobjects/AdvectiveFluxCalculator.h]: `ElementUserObject::execute()` is called on every element, that is equivalent to `ElementLoopUserObject::computeElement()` except all reinits are done for you so you do not have to deal with them. The `pre()` method would be equivalent to `ElementUserObject::initialize()`.

The difference between the 2 is that all `ElementUserObjects` are executed in one mesh sweep. While `ElementLoopUserObjects` will do 1 mesh sweep per instance.

**Comment by WilkAndy** [comment]
Hi @andrsd , i'd like ideas on how to make this latest commit better.  As you can see, i made `addJacobianBlock` of `Assembly` public, and i'm betting you guys don't like that.  I use it in `FluxLimitedTVDAdvection` to add all the Jacobian entries (next-nearest neighbours as well as nearest-neighbours)

**Comment by WilkAndy** [comment]
Hi @andrsd .  I'm back from my trip and would like to start working on this again.  Could you please comment on:

Hi @andrsd , i'd like ideas on how to make this latest commit better. As you can see, i made addJacobianBlock of Assembly public, and i'm betting you guys don't like that. I use it in FluxLimitedTVDAdvection to add all the Jacobian entries (next-nearest neighbours as well as nearest-neighbours)

**Comment by andrsd** [comment]
I will be able to get to this next week - too many other things that have higher priority.

**Comment by WilkAndy** [comment]
Cool, thanks @andrsd .   I will work on other things in the meantime.  But i don't want to leave this too long as i have to complete the whole thing - PorousFlow version - by Christmas

**Comment by WilkAndy** [comment]
Just reminding @andrsd to comment on the commit above

File: modules/porous_flow/src/kernels/FluxLimitedTVDAdvection.C
Comment (andrsd) [inline modules/porous_flow/src/kernels/FluxLimitedTVDAdvection.C]: Ad -> Add

File: modules/porous_flow/src/kernels/FluxLimitedTVDAdvection.C
Comment (WilkAndy) [reply modules/porous_flow/src/kernels/FluxLimitedTVDAdvection.C]: That works, thanks @andrsd

**Comment by WilkAndy** [comment]
Hi @andrsd .   I think it is appropriate to merge this if it passes and if you see no problems.  I have not yet got the multi-processor case working (#12557 #12558), but the PR is growing.  Next i need to do the PorousFlow bit, so now is a natural break point.

**Review by andrsd — CHANGES_REQUESTED** [review]

**Comment by WilkAndy** [comment]
Thanks for all those comments on the Kuzmin-Turek stuff, @andrsd .  It'll be a few days before i can address them, so you're off the hook for a while :-)

a

File: modules/porous_flow/src/userobjects/AdvectiveFluxCalculatorBase.C
Comment (WilkAndy) [reply modules/porous_flow/src/userobjects/AdvectiveFluxCalculatorBase.C]: I'm not sure about this, @andrsd.  I don't think the user can ever trigger this Assert by making a weirdo input file or mesh.  I think the only way of triggering the Assert is by poor C++ coding, ie, mainly the Assert is for me alone.  So i think an Assert is better than something that's going to be in the optimized code.   Do you agree?

File: modules/porous_flow/include/userobjects/AdvectiveFluxCalculatorBase.h
Comment (WilkAndy) [reply modules/porous_flow/include/userobjects/AdvectiveFluxCalculatorBase.h]: I think i do want a copy of the map in this case.

**Comment by WilkAndy** [comment]
All set to merge?  I think i've addressed all of @andrsd 's comments

File: modules/porous_flow/src/userobjects/AdvectiveFluxCalculatorBase.C
Comment (andrsd) [reply modules/porous_flow/src/userobjects/AdvectiveFluxCalculatorBase.C]: I was thinking that for the case when people (developers) will be using your API, it would be nice to guard it, because they might get into weird problems. So it is better to error out where the actual error is. We had a similar problem in RELAP and the code was bombing out in a different spot and you where not able to tell. So I like to error out as soon as possible.

However, if you do not think it is a problem, keep it this way.

File: modules/porous_flow/include/userobjects/AdvectiveFluxCalculatorBase.h
Comment (andrsd) [reply modules/porous_flow/include/userobjects/AdvectiveFluxCalculatorBase.h]: Then, you want to document it in the comment, because that will be the first thing people will ask ("why are you returning a copy").

File: modules/porous_flow/src/userobjects/AdvectiveFluxCalculatorBase.C
Comment (WilkAndy) [reply modules/porous_flow/src/userobjects/AdvectiveFluxCalculatorBase.C]: I think your experience trumps my belief

File: modules/porous_flow/include/userobjects/AdvectiveFluxCalculatorBase.h
Comment (WilkAndy) [reply modules/porous_flow/include/userobjects/AdvectiveFluxCalculatorBase.h]: I've changed this to
```
void zeroedConnection(std::map<dof_id_type, Real> & the_map, dof_id_type node_i);
```
The method clears and appropriately initializes `the_map`

**Review by andrsd — APPROVED** [review]

**Comment by WilkAndy** [comment]
Thanks for all the careful inspections and advice on this PR, @andrsd

**Pull request merged**
